lightning-invoice/src/utils.rs

```diff
@@ -5,22 +5,119 @@ use payment::{Payer, Router};
 
 use bech32::ToBase32;
 use bitcoin_hashes::Hash;
+use bitcoin_hashes::sha256::Hash as Sha256;
 use crate::prelude::*;
 use lightning::chain;
 use lightning::chain::chaininterface::{BroadcasterInterface, FeeEstimator};
 use lightning::chain::keysinterface::{Sign, KeysInterface};
 use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
-use lightning::ln::channelmanager::{ChannelDetails, ChannelManager, PaymentId, PaymentSendFailure, MIN_FINAL_CLTV_EXPIRY};
+use lightning::ln::channelmanager::{ChannelDetails, ChannelManager, PaymentId, PaymentSendFailure, MIN_FINAL_CLTV_EXPIRY, MIN_CLTV_EXPIRY_DELTA};
 use lightning::ln::msgs::LightningError;
 use lightning::routing::scoring::Score;
 use lightning::routing::network_graph::{NetworkGraph, RoutingFees};
 use lightning::routing::router::{Route, RouteHint, RouteHintHop, RouteParameters, find_route};
 use lightning::util::logger::Logger;
+use secp256k1::Secp256k1;
 use secp256k1::key::PublicKey;
 use core::convert::TryInto;
 use core::ops::Deref;
 use core::time::Duration;
 
+#[cfg(feature = "std")]
+/// Utility to create an invoice that can be paid to one of multiple nodes, or a "phantom invoice."
+/// This works because the invoice officially pays to a fake node (the "phantom"). Useful for
+/// load-balancing payments between multiple nodes.
+///
+/// `channels` contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
+/// real_channel_details)`, where `phantom_node_scid` is unique to each `channels` entry and should
+/// be retrieved freshly for each new invoice from [`ChannelManager::get_phantom_scid`].
```

Comment:
So let's say there are 3 nodes that would like to accept a single invoice, each time `create_phantom_invoice` (on a single node) is called, would the caller need to ping the other two nodes to get fresh phantom scids + usable channels to include in this `channels` list? So channel list would be a concatenated list of channels from the three nodes? If each of the 3 nodes had 2 open channels, the list size could be a size of 6?

Trying to make this as clear as possible:
```
/// `channels` contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
/// real_channel_details)`, where `phantom_node_scid` is unique to each `channels` entry and should
/// be retrieved freshly for each new invoice from [`ChannelManager::get_phantom_scid`].
/// `real_channel_details` comes from [`ChannelManager::list_channels`]. For multiple nodes,
/// `channels` will be a concatenated list of entries from each node, as each node must provide a fresh
/// `phantom_scid` and `real_channel_details` for all usable channels on each new invoice generation.
```

Reply:
Thanks, sorry the docs are pretty WIP. Will prioritize shoring them up.

> each time create_phantom_invoice (on a single node) is called, would the caller need to ping the other two nodes to get fresh phantom scids + usable channels to include in this channels list?

You don't have to retrieve fresh phantom scids on each invoice. Doing so provides more privacy, but it's optional! Fine to cache them.

If some of the provided real channels aren't usable, that could obviously result in failed payment attempts, but since channels don't close too frequently it should be fine to just poll and cache, not fetch freshly on each invoice.

> So channel list would be a concatenated list of channels from the three nodes? If each of the 3 nodes had 2 open channels, the list size could be a size of 6?

Yup! The parameters could probably be a little cleaner (suggestions welcome here) though somewhat limited by our bindings generator

Reply:
> Fine to cache them

sweet, good to know!

> Yup! The parameters could probably be a little cleaner (suggestions welcome here) though somewhat limited by our bindings generator

List of tuples makes sense to me. I think if we clarify what can be cached + the fact that the list of tuples can be formed from multiple nodes that would be good.

Maybe also a warning about including too many channels, as that might make the invoice string too long for QR codes?